vsintegration/src/FSharp.Editor/AutomaticCompletion/BraceCompletionSessionProvider.fs

```diff
@@ -0,0 +1,555 @@
+        | VerticalBar.OpenCharacter -> VerticalBarCompletionSession() :> IEditorBraceCompletionSession
+        | AngleBrackets.OpenCharacter -> AngleBracketCompletionSession() :> IEditorBraceCompletionSession
+        | DoubleQuote.OpenCharacter -> DoubleQuoteCompletionSession() :> IEditorBraceCompletionSession
+        | Asterisk.OpenCharacter -> DoubleQuoteCompletionSession() :> IEditorBraceCompletionSession
+        | _ -> null
+
+    interface IEditorBraceCompletionSessionFactory with
+
+        member this.TryCreateSession(document, openingPosition, openingBrace, cancellationToken) = 
+            if this.IsSupportedOpeningBrace(openingBrace) && this.CheckCodeContext(document, openingPosition, openingBrace, cancellationToken) then
+                this.CreateEditorSession(document, openingPosition, openingBrace, cancellationToken)
+            else
+                null
+
+[<Export(typeof<IBraceCompletionSessionProvider>)>]
+[<ContentType(FSharpConstants.FSharpContentTypeName)>]
+[<BracePair(Parenthesis.OpenCharacter, Parenthesis.CloseCharacter)>]
+[<BracePair(CurlyBrackets.OpenCharacter, CurlyBrackets.CloseCharacter)>]
+[<BracePair(SquareBrackets.OpenCharacter, SquareBrackets.CloseCharacter)>]
+[<BracePair(DoubleQuote.OpenCharacter, DoubleQuote.CloseCharacter)>]
+[<BracePair(VerticalBar.OpenCharacter, VerticalBar.CloseCharacter)>]
+[<BracePair(AngleBrackets.OpenCharacter, AngleBrackets.CloseCharacter)>]
+[<BracePair(Asterisk.OpenCharacter, Asterisk.CloseCharacter)>]
+type BraceCompletionSessionProvider
+    [<ImportingConstructor>] 
+    (
+        undoManager: ITextBufferUndoManagerProvider,
+        editorOperationsFactoryService: IEditorOperationsFactoryService
+    ) =
+
+    interface IBraceCompletionSessionProvider with
+
+        member __.TryCreateSession(textView, openingPoint, openingBrace, closingBrace, session) =
+            session <-
+                maybe {
+                    let! document =       openingPoint.Snapshot.GetOpenDocumentInCurrentContextWithChanges() |> Option.ofObj
+                    let! sessionFactory = tryGetLanguageService<IEditorBraceCompletionSessionFactory> document
+                    let! session =        sessionFactory.TryCreateSession(document, openingPoint.Position, openingBrace, CancellationToken.None) |> Option.ofObj
+
+                    let undoHistory = undoManager.GetTextBufferUndoManager(textView.TextBuffer).TextBufferUndoHistory
+                    return BraceCompletionSession(
+                        textView,
+                        openingPoint.Snapshot.TextBuffer,
+                        openingPoint, 
+                        openingBrace, 
+                        closingBrace, 
+                        undoHistory, 
+                        editorOperationsFactoryService, 
+                        session) :> IBraceCompletionSession
+                }
+                |> Option.toObj
+
+            match session with
+            | null -> false
+            | _ -> true
```

Comment:
`session <> null` maybe? I don't think the poor perf of `<> null` should play role here.